### DISCUSSION TITLE
Why do browsers perform so much worse than streamlink?

### DISCUSSION THREAD

**Discussion by TrantaLocked** [streamlink/streamlink#4855] [General]
I found that the average CPU usage of a full streamlink setup (VLC+python+Chatty with all emotes enabled) is often under 10% total, yet it tends to be much higher watching in browsers with spikes up to 30% usage, in addition to higher GPU usage too! And the CPU spikes have gotten worse for Twitch in browsers over the last couple years. 

So what is going on? Why are browsers so inefficient with Twitch? I can see that hardware acceleration is working in browsers since I can see the GPU video engine engaged in monitoring tools, it just seems vastly less efficient than streamlink.

**Comment by GabenGar** [comment]
Chatty doesn't support all emotes and the frontend for twitch does a lot more than what chatty implemented. Not to mention they run in completely different environments, the browser one being a lot more strict. The biggest difference the browser is basically a VM and everything should go through it to interact with anything on your machine, which adds general overhead to everything. On other hand vlc + streamlink combo is mostly doing raw OS calls to work with memory and GPU while chatty just showing barebones interface. I am sure chatty can be made laggier by migrating to whatever fashionable Java GUI framework right now and adding transitions and explosions everywhere.

**Reply by TrantaLocked** [reply]
> Chatty doesn't support all emotes and the frontend for twitch does a lot more than what chatty implemented. Not to mention they run in completely different environments, the browser one being a lot more strict. The biggest difference the browser is basically a VM and everything should go through it to interact with anything on your machine, which adds general overhead to everything. On other hand vlc + streamlink combo is mostly doing raw OS calls to work with memory and GPU while chatty just showing barebones interface. I am sure chatty can be made laggier by migrating to whatever fashionable Java GUI framework right now and adding transitions and explosions everywhere.

And browsers need to be like a VM for security reasons? There's no reasonable way around this?

**Reply by GabenGar** [reply]
Of course for security reasons, how else do you expect to run arbitrary 3rd party scripts without complete environment isolation? Even separate tabs run in a completely isolated fashion and only allowed to interact with each other according to a strict ruleset.
